File: Python/5kyu/PaginationHelper.py
class PaginationHelper:

    # The constructor takes in an array of items and a integer indicating
    # how many items fit within a single page
    def __init__(self, collection, items_per_page):
        self.last_page = 0
        self.items = 0
        self.pages = 0
        self.items_per_page = items_per_page

        if collection != None and len(collection) > 0:
            self.items = len(collection)

            self.last_page = 1 if self.items % items_per_page != 0 else 0
            self.pages = self.items // items_per_page + self.last_page

    # returns the number of items within the entire collection
    def item_count(self):
        return self.items

    # returns the number of pages
    def page_count(self):
        return self.pages

    # returns the number of items on the current page. page_index is zero based
    # this method should return -1 for page_index values that are out of range
    def page_item_count(self, page_index):
        if page_index < 0 or page_index >= self.page_count() or self.item_count() == 0:
            return -1

        if page_index == self.page_count() - 1:
            return self.item_count() % self.items_per_page

        return self.items_per_page

    # determines what page an item is on. Zero based indexes.
    # this method should return -1 for item_index values that are out of range
    def page_index(self, item_index):
        if item_index < 0 or item_index > self.item_count():
            return -1

        return (item_index) // self.items_per_page


collection = range(1, 25)
helper = PaginationHelper(collection, 10)

print(helper.page_count(), '===> 3')
print(helper.page_index(23), '===> 2')
print(helper.item_count(), '===> 24')

File: JavaScript/5kyu/PaginationHelper.js
// TODO: complete this object/class

// The constructor takes in an array of items and a integer indicating how many
// items fit within a single page
function PaginationHelper(collection, itemsPerPage) {
  this.collection = collection;

  this.items = collection != null ? collection.length : 0;
  this.itemsPerPage = itemsPerPage;
}

// returns the number of items within the entire collection
PaginationHelper.prototype.itemCount = function () {
  return this.items;
};

// returns the number of pages
PaginationHelper.prototype.pageCount = function () {
  return Math.ceil(this.items / this.itemsPerPage);
};

// returns the number of items on the current page. page_index is zero based.
// this method should return -1 for pageIndex values that are out of range
PaginationHelper.prototype.pageItemCount = function (pageIndex) {
  if (pageIndex > this.pageCount() || pageIndex < 0) return -1;
  if (pageIndex + 1 == this.pageCount()) return this.items % this.itemsPerPage;

  return this.itemsPerPage;
};

// determines what page an item is on. Zero based indexes
// this method should return -1 for itemIndex values that are out of range
PaginationHelper.prototype.pageIndex = function (itemIndex) {
  if (itemIndex > this.itemCount() || itemIndex < 0) return -1;

  return Math.floor((itemIndex + 1) / this.itemsPerPage);
};

var helper = new PaginationHelper(["a", "b", "c", "d", "e", "f"], 4);
console.log(helper.pageCount()); //should == 2
console.log(helper.itemCount()); //should == 6
console.log(helper.pageItemCount(0)); //should == 4
console.log(helper.pageItemCount(1)); // last page - should == 2
console.log(helper.pageItemCount(2)); // should == -1 since the page is invalid

// pageIndex takes an item index and returns the page that it belongs on
console.log(helper.pageIndex(5)); //should == 1 (zero based index)
console.log(helper.pageIndex(2)); //should == 0
console.log(helper.pageIndex(20)); //should == -1
console.log(helper.pageIndex(-10)); //should == -1
console.log(helper.pageIndex(0)); //should == 0
